Support whitelist of dynamic sources

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147979

This PR introduces the ability to whitelist sources as dynamic. This is particularly useful for large models with graph breaks, as you can keep the dynamism across graph breaks since source names stay consistent. Additionally you can use this to mark ints as dynamic.

NB: I intentionally didn't complicate the interface by supporting specification of per dimension dynamism. There is virtue in keeping true to the standard way of representing sources (eg. L['x']). If we find in practice that we need more more fine grained control, we can explore further affordances at that time.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
Support whitelist of dynamic sources

[ghstack-poisoned]

Mingming-Ding reviewed (approved):
LGTM

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2639,6 +2643,15 @@ def get_automatic_dynamic_shapes_mark_as():
         )
 
 
+def get_dynamic_sources() -> set[str]:
```
nit: maybe just define put _DYNAMIC_SOURCES on top of this function. since its only ever used by this function

laithsakka commented on torch/compiler/config.py:
```diff
@@ -63,5 +63,13 @@
 A common use case for such a tag is to break caches.
 """
 
+dynamic_sources: str = Config(
```
can we clearly define and document the precedence of test over dynamic = False for example and over config.assume_static_by_default ..
and over assume params static and assume nn module attributes static and add tests to guarantee the ordering

bobrenjc93 replied:
Yeah let me do that in a separate PR that I'll stack on top of this one.

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2807,7 +2821,7 @@ def update_dim2constraint(dim, constraint_range, name):
         # Reflect the user directive in the frame_state
         # For dynamic, apply None always
-        if marked_dynamic:
+        if marked_dynamic or source.name() in dynamic_sources:
```
the log bellow 
log.debug("automatic dynamic %s marked dynamic", name) would be confusing .mmm thinking

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -7858,6 +7858,54 @@ def my_dyn_fn(x, y):
             torch.compile(my_dyn_fn, backend="eager")(y, y)
 
+    @torch.compiler.config.patch(dynamic_sources="L['x']")
+    def test_dynamic_int_sources(self):
```
nit: if you same same prefix for all we can test them and find them in one command 
test_dynamic_sources_int/tesnor..etc
then I can do -k test_dynamic_sources

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1955,7 +1956,10 @@ def wrap_symint(self, value):
-            elif not config.assume_static_by_default:
+            elif (
+                not config.assume_static_by_default
+                or self.source.name() in get_dynamic_sources()
```
```
   if (self.source.name() in get_dynamic_sources()):
                 dynamic_dim = DimDynamic.DYNAMIC
            elif (
                config.automatic_dynamic_shapes
                and frame_state_entry.scalar is auto_dynamic
            ):
                dynamic_dim = get_automatic_dynamic_shapes_mark_as()
            elif (
                not config.assume_static_by_default
            ):
                dynamic_dim = DimDynamic.DYNAMIC
  ```
This way you do not need the changes later for automatic dynamic

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1955,7 +1956,10 @@ def wrap_symint(self, value):
+            elif (
+                not config.assume_static_by_default
+                or self.source.name() in get_dynamic_sources()
+            ):
```
can we add a log
marked dynamic due to XX

laithsakka reviewed (dismissed):
looks great some not major comments.
in the future if needed we can add regex support easily on top of this. 

this is very nice

bobrenjc93 committed:
Update on "Support whitelist of dynamic sources"


This PR introduces the ability to whitelist sources as dynamic. This is particularly useful for large models with graph breaks, as you can keep the dynamism across graph breaks since source names stay consistent. Additionally you can use this to mark ints as dynamic.

NB: I intentionally didn't complicate the interface by supporting specification of per dimension dynamism. There is virtue in keeping true to the standard way of representing sources (eg. L['x']). If we find in practice that we need more more fine grained control, we can explore further affordances at that time.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Support whitelist of dynamic sources"


This PR introduces the ability to whitelist sources as dynamic. This is particularly useful for large models with graph breaks, as you can keep the dynamism across graph breaks since source names stay consistent. Additionally you can use this to mark ints as dynamic.

NB: I intentionally didn't complicate the interface by supporting specification of per dimension dynamism. There is virtue in keeping true to the standard way of representing sources (eg. L['x']). If we find in practice that we need more more fine grained control, we can explore further affordances at that time.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Support whitelist of dynamic sources"


This PR introduces the ability to whitelist sources as dynamic. This is particularly useful for large models with graph breaks, as you can keep the dynamism across graph breaks since source names stay consistent. Additionally you can use this to mark ints as dynamic.

NB: I intentionally didn't complicate the interface by supporting specification of per dimension dynamism. There is virtue in keeping true to the standard way of representing sources (eg. L['x']). If we find in practice that we need more more fine grained control, we can explore further affordances at that time.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
